### PR TITLE
Fix precaching to be consistent

### LINKED ISSUE #75: sw-precache does not include all entry points
It looks like this happens only when generating a fresh build and not when a previous build already exists.

### PR DESCRIPTION
**Pull request by azakus**
sw-precache reads directly from disk to generate precache manifests, so we
must wait for the streams to be complete before calling
generateServiceWorker.

Read the config file once outside of the stream, let sw-precache write to
disk

Clone config so that bundled and unbundled don't write into each other

Add tests for service worker generation

Also handle uncaught promises in build

Fixes #75

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,6 +55,7 @@
     "eslint": "^2.9.0",
     "gulp-mocha": "^2.2.0",
     "sinon": "^1.17.4",
+    "tmp": "0.0.28",
```
Comment (justinfagnani) [inline package.json]: what's the difference between this an `temp`? @garlicnation is adding `temp` in #46

**Comment by azakus** [comment]
@justinfagnani PTAL

File: src/build/build.ts [hunk 2]
```diff
@@ -45,8 +46,13 @@ process.on('uncaughtException', (err) => {
   console.error(err.stack);
 });
 
+process.on('unhandledRejection', (err, p) => {
+  console.log(`Promise rejection: ${err}`);
```
Comment (justinfagnani) [inline src/build/build.ts]: maybe add `console.error(err.stack)`

File: src/build/build.ts [hunk 3]
```diff
@@ -45,8 +46,13 @@ process.on('uncaughtException', (err) => {
   console.error(err.stack);
 });
 
+process.on('unhandledRejection', (err, p) => {
```
Comment (justinfagnani) [inline src/build/build.ts]: I like longer names, like `error` and `promise`. We only have to type them once, then read them foreverrrr...

File: src/build/build.ts [hunk 4]
```diff
@@ -108,50 +114,49 @@ export function build(options?: BuildOptions): Promise<any> {
...
+      let unbundledDeps = Array.from(bundler.streamResolver.requestedUrls);
+
+      let bundledDeps = Array.from(bundler.entrypointFiles.keys());
+      bundledDeps.push(bundler.sharedBundleUrl);
+
+      parsePreCacheConfig(swPrecacheConfig).then((swConfig) => {
+        let unbundledConfig = clone(swConfig);
+        let bundledConfig = clone(swConfig);
+        Promise.all([
+          generateServiceWorker({
```
Comment (justinfagnani) [inline src/build/build.ts]: how about making a local wrapper on `generateServiceWorker` that takes care of the common params?

**Comment by justinfagnani** [comment]
big LGTM

**Pull request merged**